### PR TITLE
ACS-6379: [HxI LI] S3 Credentials - PART I

### PR DESCRIPTION
**Pull request by krdabrowski**
This PR only adds part of the functionality:
- adding component allowing to upload objects to S3 storage using preSigned URL
- adding integration test based on Testcontainers and LocalStack
- fixed Maven and test config allowing integration tests to run using properties from pom file

what is still missing and will be added in next part:
- way to call HxI API, which will return preSigned URL

For more see: https://alfresco.atlassian.net/browse/ACS-6379

### CODE REVIEW

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+import org.testcontainers.utility.DockerImageName;
+
+import org.alfresco.hxi_connector.live_ingester.storage.local.LocalStorageClient;
+import org.alfresco.hxi_connector.live_ingester.storage.local.LocalStorageConfig;
+import org.alfresco.hxi_connector.live_ingester.utils.DockerTags;
+
+@SpringBootTest(classes = {LocalStorageConfig.class, SignedStorageClient.class})
+@ActiveProfiles({"test"})
+@Testcontainers
+class StorageClientIntegrationTest
```
Comment (Pawel-608) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: I wonder if we could separate unit and integration tests by adding a new test source, like here: https://github.com/Alfresco/hxi-connector/tree/feature/add-it-test. It is really easy to configure and can be really handy as you don't mix unrelated stuff. wdyt @krdabrowski @tpage-alfresco @mpichura?

Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: I was also thinking about that and so far, I don't see any contraindications 🤔 I can adjust it, but let see what rest will say

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+import lombok.AccessLevel;
+import lombok.NoArgsConstructor;
+import lombok.SneakyThrows;
+
+/**
+ * Allows to access properties from 'docker-tags.properties' file. It's required to run: `mvn package` before accessing the properties file, to allow Maven filter out and replace variables.
+ */
+@NoArgsConstructor(access = AccessLevel.PRIVATE)
+public class DockerTags
```
Comment (Pawel-608) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:41]: There are two test utils (`utils` and `util`) packages now ;) (probably because of rebase) I would move everything to one of them and remove other one

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        return Optional.ofNullable(properties).map(p -> p.getProperty(key)).orElse(null);
+    }
+
+    public static String getOrDefault(String propertyKey, String defaultValue)
+    {
+        if (properties == null)
+        {
+            loadProperties();
```
Comment (Pawel-608) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:62]: I am wondering if maybe we could use just `getProperty()` that will throw exception if property is not found and also make `loadProperties()` throw an error if properties doesn't exists as we probably expect that it should always be available and this approach will allow us to fail fast

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    private StorageClient storageClient;
+
+    @BeforeAll
+    static void beforeAll() throws IOException, InterruptedException
+    {
+        localStack.execInContainer("awslocal", "s3api", "create-bucket", "--bucket", BUCKET_NAME);
+    }
+
+    @Test
+    void testUpload() throws IOException
```
Comment (Pawel-608) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:81]: maybe it is worth to add @SneakyThrows?

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        List<String> freshBucketContent = testS3Storage.listBucketContent(BUCKET_NAME);
+        assertThat(freshBucketContent)
+                .isNotNull()
+                .isEmpty();
+
+        // given
+        try (InputStream fileInputStream = new ByteArrayInputStream(OBJECT_CONTENT.getBytes()))
```
Comment (Pawel-608) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:89]: I am not sure as I never used it but when I was checking lombok features I found this annotation https://projectlombok.org/features/Cleanup, dyt we can use it here?

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (tpage-alfresco) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: Separating `src/test/java` from `src/integration-test/java` is fine. Doing a quick scan of our existing projects then it seems more standard to rename the other way - into `src/unit-test/java` and `src/test/java`.

File: distribution/pom.xml [hunk 6]
```diff
@@ -73,7 +73,7 @@
                     <plugin>
                         <groupId>org.apache.maven.plugins</groupId>
                         <artifactId>maven-assembly-plugin</artifactId>
-                        <version>3.6.0</version>
+                        <version>${maven-assembly-plugin.version}</version>
```
Comment (tpage-alfresco) [inline distribution/pom.xml:76]: I guess possibly even better is to add a pluginManagement section if we want to manage this centrally.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClient.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+ */
+package org.alfresco.hxi_connector.live_ingester.storage;
+
+import java.io.File;
+import java.io.InputStream;
+import java.net.URL;
+
+import org.apache.http.StatusLine;
+
+public interface StorageClient
```
Comment (Pawel-608) [inline live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClient.java:34]: I think we should move this interface to domain and probably adjust it a little as it shows internal implementation details (use of apache and pre-signed urls) but it is probably something we can do later (I have no idea how we want to use it so idk how we can improve it😂)

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+            httpPut.setEntity(entity);
+            httpPut.setHeader(CONTENT_TYPE, contentType);
+            // TODO If additional metadata will be required uncomment bellow, otherwise removed it
+            // metadata.forEach((k, v) -> httpPut.setHeader("x-amz-meta-" + k, v));
+
+            HttpResponse response = httpClient.execute(httpPut);
+            return response.getStatusLine();
+        } catch (URISyntaxException e)
+        {
+            throw new IllegalArgumentException("Pre Signed URL cannot be parsed to URI", e);
```
Comment (Pawel-608) [inline live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java:82]: I would stick to `LiveIngesterRuntimeException` and its subclasses (you can always create custom one)

File: live-ingester/pom.xml [hunk 9]
```diff
@@ -53,6 +53,22 @@
 			<artifactId>acs-event-model</artifactId>
 		</dependency>
 
+		<dependency>
+			<groupId>org.testcontainers</groupId>
+			<artifactId>junit-jupiter</artifactId>
+			<scope>test</scope>
```
Comment (tpage-alfresco) [inline live-ingester/pom.xml:59]: It's probably fairly safe to put the scope of these three dependencies in the `dependencyManagement` section alongside the version.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        try (CloseableHttpClient httpClient = HttpClients.createDefault())
+        {
+            HttpPut httpPut = new HttpPut(preSignedUrl.toURI());
+            HttpEntity entity = EntityBuilder.create()
+                    .setStream(inputStream)
+                    .build();
+            httpPut.setEntity(entity);
+            httpPut.setHeader(CONTENT_TYPE, contentType);
+            // TODO If additional metadata will be required uncomment bellow, otherwise removed it
+            // metadata.forEach((k, v) -> httpPut.setHeader("x-amz-meta-" + k, v));
```
Comment (Pawel-608) [inline live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java:76]: I would delete this comment

**Review by Pawel-608 — APPROVED** [review]
generally LGTM, the only thing is that we probably will need to adjust `StorageClient` interface at some point

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+    private static final String LOCALSTACK_IMAGE = "localstack/localstack";
+    private static final String LOCALSTACK_TAG = DockerTags.getOrDefault("localstack.tag", "3.0.2");
+    private static final String BUCKET_NAME = "test-hxinsight-bucket";
+    private static final String OBJECT_KEY = "dummy.txt";
+    private static final String OBJECT_CONTENT = "dummy's file dummy content";
+    private static final String OBJECT_CONTENT_TYPE = "plain/text";
+
+    @Container
+    @SuppressWarnings("PMD.FieldNamingConventions")
+    private static final LocalStackContainer localStack = new LocalStackContainer(DockerImageName.parse(LOCALSTACK_IMAGE).withTag(LOCALSTACK_TAG));
```
Comment (tpage-alfresco) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:67]: Why didn't you want to call this `LOCAL_STACK`? Maybe you could add a comment to explain the suppression if this has to match with somewhere else in the code?

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (Pawel-608) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: can we separate it then?

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+    {
+        localStack.execInContainer("awslocal", "s3api", "create-bucket", "--bucket", BUCKET_NAME);
+    }
+
+    @Test
+    void testUpload() throws IOException
+    {
+        List<String> freshBucketContent = testS3Storage.listBucketContent(BUCKET_NAME);
+        assertThat(freshBucketContent)
+                .isNotNull()
```
Comment (tpage-alfresco) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:85]: I don't think you need this check - `isEmpty` already covers it.

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+            // when
+            int actualStatusCode = storageClient.upload(fileInputStream, OBJECT_CONTENT_TYPE, preSignedUrl).getStatusCode();
+
+            // then
+            assertThat(actualStatusCode).isEqualTo(200);
+            List<String> actualBucketContent = testS3Storage.listBucketContent(BUCKET_NAME);
+            assertThat(actualBucketContent)
+                    .isNotNull()
+                    .containsOnly(OBJECT_KEY);
```
Comment (tpage-alfresco) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:101]: I don't think you need the `isNotNull` here, and I think it would be more accurate to replace `containsOnly` with `isEqualTo` (since otherwise `[OBJECT_KEY, OBJECT_KEY]` will pass the test).

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (mpichura) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: It's definitely worth separating out different types of tests.
Although I usually used to do that with packages (unit tests in same packages as tested classes, integration tests in separate package), I am okay with separate test sources as this at times be even more effective.

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+        }
+
+        String value = defaultValue;
+        if (properties != null)
+        {
+            String property = properties.getProperty(propertyKey);
+            if (property != null && !property.startsWith("@") && !property.endsWith("@"))
```
Comment (tpage-alfresco) [inline live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:69]: I'm not sure I understand the idea behind the second part of this assertion - if a property is not substituted properly by maven then we use the default value? I think this would lead to hard to track down issues - wouldn't it be easier if we just got the `@something@` value instead and then saw this mentioned in the resulting error message?

**Review by tpage-alfresco — COMMENTED** [review]
Overall this looks great! I tried out the StorageClientIntegrationTest locally and it was really fast. Thanks Krys!

File: distribution/pom.xml (hunk 6 above)
Comment (krdabrowski) [reply distribution/pom.xml:76]: I will do it with pleasure

File: live-ingester/pom.xml (hunk 9 above)
Comment (krdabrowski) [reply live-ingester/pom.xml:59]: In the other hand in future it still will be possible to to remove the `test` scope from `dependencyManagement` section

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java (hunk 2 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:41]: Didn't notice that - I will fix it, thanks

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 4 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:81]: Over test methods it doesn't bring any value IMO

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: Ok, I will work on that - I will go with `src/test/java` and `src/integration-test/java` as it's more popular convention in Java world. Also I think, that unit test are these primary kind of tests and they shouldn't be offended with some prefixes/suffixes.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClient.java (hunk 7 above)
Comment (krdabrowski) [reply live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClient.java:34]: So this the issue at this stage - I'm not sure how it will look finally. It will clarify in the second part of this task. In general it's a representation of an service/component, so I'm not sure if `domain` is the right place. Overall I would leave it as it is, till the second part is done.

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java (hunk 10 above)
Comment (krdabrowski) [reply live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java:76]: I'm not sure if it won't be necessary in the second part, for now I would leave it.

Comment (Pawel-608) [reply live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java:76]: Are you sure you won't forget to remove it if you don't need it?😂

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 11 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:67]: Yeah, so there is a live debate in Java society about what should be written using SNAKE_CASE an what not. I believe that only constants should be written this way. What is a constant is very debatable, there is quite cool explanation here: https://stackoverflow.com/a/12069687/2211974 - I like it 😅 Overall IMO `localStack` is a final reference not a CONSTANT, so I called it using camelCase.

Comment (tpage-alfresco) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:67]: Ah - yes, I read that before. However since we've all had years of seeing `CAPS == static final` then the discussion seemed a bit moot. I think the convenience of using the `@Slf4j` annotation is worth an exception however. Overall I don't mind either way, but if we want to use this style more widely we should probably remove the PMD check.

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java (hunk 3 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:62]: This is true, but I must admit, that it's quite annoying when you open a project, try to run a test and it fails because you need to make some step (e.g. `mvn package`) earlier. I like tests which run smoothly without any prerequisites. This is the reason why there are two methods: `getProperty()` and `getOrDefault()`, first one in did should fail if file is missing, but the second one shouldn't care if the file or property exists - if not return default value.
In the other hand there is this small risk, that even on CI something could go wrong with Docker image tag substitution because of what default value would be picked up and we wouldn't notice it. At the same time it's hard for me to image a situation where Maven would not handle the resource filtering correctly while building the app - something would have to be really messed up in config... and even if, would it be so bad to use the default value instead (even outdated)? We need to choose between convenience and reliability - I like when something works straight away, but in this case - having only one prerequisite - I could live with that. Tell me guys @Pawel-608 @tpage-alfresco @mpichura WDYT

File: live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java (hunk 10 above)
Comment (krdabrowski) [reply live-ingester/src/main/java/org/alfresco/hxi_connector/live_ingester/storage/SignedStorageClient.java:76]: I'm sure 😅

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: After small investigation it turned out, that it's not a trivial stuff to do - Maven is designed this way, that all test should be placed in one folder - of course there exist a way to workaround it, but not without issues. In general the community recommendations are: leave all kinds of tests in one (dedicated by Maven) folder, or create a separate Maven module for that: e.g. https://stackoverflow.com/questions/37998040/where-should-the-integration-tests-be-stored-when-using-the-maven-failsafe-plugi or https://stackoverflow.com/questions/46831999/maven-add-integration-tests and plenty more...
I think this issue deserves a separate Jira task, so won't be solved here. FYI @Pawel-608 @tpage-alfresco @mpichura

Comment (tpage-alfresco) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:56]: Definitely fine to raise a separate ticket for this to be done. I found several projects with `src/unit-test/java` so I think we can probably copy something from one of those to set the configuration.

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java (hunk 3 above)
Comment (tpage-alfresco) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/utils/DockerTags.java:62]: As we discussed we could:
* Do what you've done for all our "real" integration tests
* Add one test that only runs on CI which checks there are no `@something@` properties

We want to add a dev build profile anyway as there are some things that take a while and we don't need to do for each local build (e.g. download the license texts).

**Comment by krdabrowski** [comment]
@tpage-alfresco @Pawel-608 I've updated most of the stuff you mentioned.
The test mentioned here: https://github.com/Alfresco/hxi-connector/pull/20#discussion_r1425627172 I will add in PART II

File: live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java (hunk 1 above)
Comment (krdabrowski) [reply live-ingester/src/test/java/org/alfresco/hxi_connector/live_ingester/storage/StorageClientIntegrationTest.java:58]: Task created: https://alfresco.atlassian.net/browse/ACS-6514
